be1-go/hub/base_hub.go

```diff
@@ -0,0 +1,117 @@
+package hub
+
+import (
+	"fmt"
+	"log"
+	"path/filepath"
+	"sync"
+
+	"student20_pop/message"
+
+	"github.com/xeipuuv/gojsonschema"
+	"go.dedis.ch/kyber/v3"
+	"golang.org/x/xerrors"
+)
+
+type baseHub struct {
+	messageChan chan IncomingMessage
+
+	sync.RWMutex
+	channelByID map[string]Channel
+
+	public kyber.Point
+
+	schemas map[string]*gojsonschema.Schema
+}
+
+const (
+	GenericMsgSchema string = "genericMsgSchema"
+	DataSchema       string = "dataSchema"
+)
+
+// NewBaseHub returns a Base Hub.
+func NewBaseHub(public kyber.Point) (*baseHub, error) {
+	// Import the Json schemas defined in the protocol section
+	protocolPath, err := filepath.Abs("../protocol")
```

Comment:
I reckon the validation can be isolated in a separate package.

```go
type struct SchemaValidator {
    schemas map[string]*gojson.Schema
}

// RegisterSchema adds a JSON schema specified by the path
func (s *SchemaValidator) RegisterSchema(path string) error { .. }

func (s *SchemaValidator) VerifyJSON(byteMessage []byte, schema string) error { ..}

func NewSchemaValidator() (*SchemaValidator, error) {
    return NewSchemaValidatorWithSchemas("path/to/genericMessage.json", "path/to/data.json")
}

func NewSchemaValidatorWithSchemas(paths ...string) (*SchemaValidator, error) {
    // instantiate schema and register the paths using RegisterPath
}
```

The baseHub can then use an instance of a SchemaValidator. This way we decouple schema validation from hub and can test the validations independently. Does that sound like a good idea?

Reply:
Yes good idea! I implemented it. I just did a small modification. Instead of using `path string` for the agument in RegisterSchema I used a `schema struct`:
```
type schema struct {
	name string
	path string
}
```
I did that to be able to pass the `name` and the `path` of a schema to the function `RegisterSchema` and `NewSchemaValidatorWithSchemas`

Reply:
Do you think it is a good idea?